crates/core/src/factories/core_fetch.rs

```diff
@@ -0,0 +1,341 @@
+//!
+//! - Exposes public method [`CoreFetch::add_ops`] that takes a list of ops and an agent id.
+//! - Stores pairs of ([`OpId`][`AgentId`]) in the order that ops have been added.
+//!
+//! ### Fetch task
+//!
+//! - Toeks pairs of ([`OpId`], [`AgentId`]) from the beginning of the queue and requests the op from the agent.
+//! - Appends removed pair to the end of the queue.
+//! - Put unresponsive peers on a cool-down list.
+//!
+//! ### Incoming op task
+//!
+//! - Incoming op is written to the data store.
+//! - Once persisted successfully, op is removed from the data object.
+
+use std::{
+    collections::HashMap,
+    sync::Arc,
+    time::{Duration, Instant},
+};
+
+use indexmap::IndexMap;
+use kitsune2_api::{
+    builder,
+    config::ModConfig,
+    fetch::{DynFetch, DynFetchFactory, Fetch, FetchFactory},
+    tx::Transport,
+    AgentId, BoxFut, K2Result, OpId,
+};
+use tokio::{
+    select,
+    sync::{mpsc::Sender, Mutex},
+};
+
+const MOD_NAME: &str = "Fetch";
+
+/// Configuration parameters for [CoreFetchFactory].
+#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct CoreFetchConfig {
+    /// How many parallel op fetch requests can be made at once. Default: 2.  
+    parallel_request_count: u8,
+    /// Duration in ms to pause when parallel request count is reached. Default: 10.
+    parallel_request_pause: u64,
+    /// Duration in ms to sleep when idle. Default: 1_000.
+    fetch_loop_sleep: u64,
+    /// Duration in ms to keep an unresponsive agent on the cool-down list. Default: 10_000.
+    cool_down_interval: u64,
+}
+
+impl Default for CoreFetchConfig {
+    fn default() -> Self {
+        Self {
+            parallel_request_count: 2,
+            parallel_request_pause: 10,
+            fetch_loop_sleep: 1_000,
+            cool_down_interval: 10_000,
+        }
+    }
+}
+
+impl ModConfig for CoreFetchConfig {}
+
+impl CoreFetchConfig {
+    /// Get parallel request count.
+    pub fn parallel_request_count(&self) -> u8 {
+        self.parallel_request_count
+    }
+    /// Get parallel request pause interval.
+    pub fn parallel_request_pause(&self) -> u64 {
+        self.parallel_request_pause
+    }
+    /// Get fetch loop pause interval.
+    pub fn fetch_loop_pause(&self) -> u64 {
+        self.fetch_loop_sleep
+    }
+    /// Get cool-down interval.
+    pub fn cool_down_interval(&self) -> u64 {
+        self.cool_down_interval
+    }
+}
+
+// TODO: Temporary trait object of a transport module to facilitate unit tests.
+type DynTransport = Arc<Mutex<dyn Transport>>;
+
+#[derive(Debug)]
+struct CoreFetch(Inner);
+
+impl CoreFetch {
+    fn new(config: CoreFetchConfig, transport: DynTransport) -> Self {
+        Self(Inner::spawn_fetch_task(config, transport))
+    }
+}
+
+impl Fetch for CoreFetch {
+    fn add_ops(
+        &mut self,
+        op_list: Vec<OpId>,
+        source: AgentId,
+    ) -> BoxFut<'_, K2Result<()>> {
+        Box::pin(async move {
```

Comment:
Is this different to doing 

```rust
async move {
     // do stuff
}.boxed()
```

? I slightly prefer the less syntax but it'd be nice to be consistent with whichever we go with

Reply:
They are identical. I've been preferring the Box::pin version because it doesn't require an external library, it's built into rust.